### PR TITLE
Update Config Loading

### PR DESCRIPTION
**Pull request by blowmage**
WIP PR to demonstrate moving the configuration defaults to the `google-cloud-[language].rb` files already used as a loading mechanism.

### CODE REVIEW

**Comment by blowmage** [comment]
@dazuma What do you think of this approach?

**Comment by blowmage** [comment]
@dazuma Ping.

**Comment by dazuma** [comment]
I like the approach. It looks like you're creating a "lazy" or "deferred" value type that is evaluated just-in-time whenever the value is requested—not just for defaults but for _any_ value.

One possible issue is that it is no longer possible for a config value itself to be a `Proc` (which is a use case we have on the opencensus side, allowing the setting of formatters and other "strategy pattern" objects). Maybe consider creating a separate class for this type, and provide an easy-to-use factory for it, e.g.:

```ruby
class Google::Cloud::Config
  class DeferredValue
    def initialize &block
      @callback = block
    end
    def call
      @callback.call
    end
  end
  def self.deferred &block
    DeferredValue.new &block
  end
end
```

Then you can do the following, which is more explicit that special behavior is being invoked:

```ruby
config.project_id = Google::Cloud::Config.deferred { ENV["PROJECT_ID"] }
```

Also thumbs-up on setting up the configs in the "gem name" file so it's available once bundler loads the gem. I hadn't thought of that approach.

**Comment by blowmage** [comment]
> One possible issue is that it is no longer possible for a config value itself to be a `Proc`

You still can, actually. I have to do this in a couple places where `Proc` is supported in some other libraries. The fix is to have a `Proc` return the `Proc` you want to use.

```ruby
generator = -> { rand }
config.random_number_generator = -> { generator }

config.random_number_generator #=> #<Proc:0x0123 (lambda)>
````

I'll get the other libraries changed over to use the gem name file.

**Comment by dazuma** [comment]
> The fix is to have a Proc return the Proc you want to use.

Oh ick. I guess that technically works, but I'm not a fan. It means we'd be requiring users to wrap their procs in a proc as well, if they want to set a proc-valued config. That's going to cause lots of bugs when users forget to do so.

**Comment by blowmage** [comment]
Ah, I didn't realize that we already had configuration values returning a `Proc` object. We can't call the proc object since it would be a change in behavior. I will add your `DeferredValue` code instead.

**Comment by dazuma** [comment]
Actually I don't *think* we already have any Proc-valued configs here (yet). (We do in opencensus, but that isn't sharing the same config mechanism.) Did you find one here? I just thought it was likely we would in the future.

**Comment by dazuma** [comment]
Oh never mind, looks like we do have some in Trace. Okay, then yes, we need a wrapper. :)

**Comment by blowmage** [comment]
@dazuma Yeah, I got all the way to Trace before I saw that! :)

This PR is updated with the new code. PTAL

File: google-cloud-core/lib/google/cloud.rb [hunk 1]
```diff
@@ -90,36 +90,31 @@ def self.new project_id = nil, credentials = nil, retries: nil, timeout: nil
     #     Google::Cloud libraries.
     #
     def self.configure
+      @config ||= Config.create
```
Comment (dazuma) [inline google-cloud-core/lib/google/cloud.rb]: Can we just call this once during the class definition, or are you concerned about the file being required multiple times?

**Comment by blowmage** [comment]
> Are you done, or is there something else you wanted to do?

I don't have any other changes.

File: google-cloud-core/lib/google/cloud.rb (hunk 1 above)
Comment (blowmage) [reply google-cloud-core/lib/google/cloud.rb]: Its more about the variable not being used by anything other than the method, and a desire to have the variable "contained" to the method. The `configure` method isn't the inner-most loop so I don't sweat the extra nil check in regards to performance.

**Pull request merged**